util/types.go

```diff
@@ -59,264 +56,242 @@ type LeafData struct {
 		}
+		// fmt.Printf("h %d leaf %d %s len %d\n",
+		// ud.Height, i, ld.Outpoint.String(), len(ld.PkScript))
 	}
 
-	if buffer.Len() > 0 {
-		// read the TTL values, there are as many TTLs as there are targets in the proof.
-		ud.LeafTTLs = make([]uint32, len(ud.AccProof.Targets))
-		for i, _ := range ud.LeafTTLs {
-			// each ttl is 4 bytes (uint32)
-			ud.LeafTTLs[i] = binary.BigEndian.Uint32(buffer.Next(4))
+	return
+}
+
+//
+func (ud *UData) SerializeSize() int {
+	var ldsize int
+	var b bytes.Buffer
+
+	// TODO this is slow, can remove double checking once it works reliably
+	for _, l := range ud.Stxos {
+		ldsize += l.SerializeSize()
+		b.Reset()
+		l.Serialize(&b)
+		if b.Len() != l.SerializeSize() {
+			fmt.Printf(" b.Len() %d, l.SerializeSize() %d\n",
+				b.Len(), l.SerializeSize())
 		}
 	}
 
-	return ud, nil
-}
-
-func (ub *UBlock) FromBytes(argbytes []byte) (err error) {
-	buf := bytes.NewBuffer(argbytes)
-	// first deser the block, then the udata
-	err = ub.Block.Deserialize(buf)
-	if err != nil {
-		return
+	b.Reset()
+	ud.AccProof.Serialize(&b)
+	if b.Len() != ud.AccProof.SerializeSize() {
+		fmt.Printf(" b.Len() %d, AccProof.SerializeSize() %d\n",
+			b.Len(), ud.AccProof.SerializeSize())
 	}
-	ub.ExtraData, err = UDataFromBytes(buf.Bytes())
-	return
+
+	guess := 8 + (4 * len(ud.TxoTTLs)) + ud.AccProof.SerializeSize() + ldsize
+
+	// 8B height & numTTLs, 4B per TTL, accProof size, leaf sizes
+	return guess
 }
 
-// network serialization for UBlocks (regular block with udata)
-// Firstjust a wire.MsgBlock with the regular serialization.
-// Then there's  4 bytes is (big endian) length of the udata.
-// So basically a block then udata, that's it.
-// Looks like "height" doesn't get sent over this way, but maybe that's OK.
-func (ub *UBlock) Deserialize(r io.Reader) (err error) {
+func (ud *UData) Deserialize(r io.Reader) (err error) {
 
-	// first deser the block
-	err = ub.Block.Deserialize(r)
-	if err != nil {
+	err = binary.Read(r, binary.BigEndian, &ud.Height)
+	if err != nil { // ^ 4B block height
+		fmt.Printf("ud deser Height err %s\n", err.Error())
 		return
 	}
+	// fmt.Printf("read height %d\n", ud.Height)
 
-	// fmt.Printf("deser block OK %s\n", ub.Block.Header.BlockHash().String())
-	var uDataLen, bytesRead uint32
-	var n int
-	// read udata length
-	err = binary.Read(r, binary.BigEndian, &uDataLen)
-	if err != nil {
+	var numTTLs int32
+	err = binary.Read(r, binary.BigEndian, &numTTLs)
+	if err != nil { // ^ 4B num ttls
+		fmt.Printf("ud deser numTTLs err %s\n", err.Error())
 		return
 	}
-	// fmt.Printf("server says %d byte uDataLen\n", uDataLen)
-
-	udataBytes := make([]byte, uDataLen)
+	// fmt.Printf("read ttls %d\n", numTTLs)
+	// fmt.Printf("UData deser read h %d - %d ttls ", ud.Height, numTTLs)
 
-	for bytesRead < uDataLen {
-		n, err = r.Read(udataBytes[bytesRead:])
+	ud.TxoTTLs = make([]int32, numTTLs)
+	for i, _ := range ud.TxoTTLs { // write all ttls
+		err = binary.Read(r, binary.BigEndian, ud.TxoTTLs[i])
```

Comment:
the ttl has to be passed as a pointer:
```suggestion
		err = binary.Read(r, binary.BigEndian, &ud.TxoTTLs[i])
```

it would also be possible to pass the slice directly:
https://github.com/golang/go/blob/d9725f549f48ae6e4eaf70311f6827173453935b/src/encoding/binary/binary.go#L210

Reply:
Oh man missed that.  And somehow that only causes it to fail intermittently / rarely?  Surprising